omnigibson/object_states/attached_to.py

```diff
@@ -0,0 +1,293 @@
+            if self.parent == other:
+                # Already attached to this object. Do nothing.
+                return True
+            elif self.parent is None:
+                # Find attachment links that satisfy the proximity requirements
+                child_link, parent_link = self._find_attachment_links(other, bypass_alignment_checking)
+                if child_link is not None:
+                    self._attach(other, child_link, parent_link)
+                    return True
+                else:
+                    return False
+            else:
+                log.debug(f"Trying to attach object {self.obj.name} to object {other.name},"
+                          f"but it is already attached to object {self.parent.name}. Try detaching first.")
+                return False
+
+        # Attempt to detach
+        else:
+            if self.parent == other:
+                self._detach()
+
+                # Wake up objects so that passive forces like gravity can be applied.
+                self.obj.wake()
+                other.wake()
+            return True
+
+    def _get_value(self, other):
+        # Simply return if the current parent matches other
+        return other == self.parent
+
+    def _find_attachment_links(self,
+                               other,
+                               bypass_alignment_checking=False,
+                               pos_thresh=m.DEFAULT_POSITION_THRESHOLD,
+                               orn_thresh=m.DEFAULT_ORIENTATION_THRESHOLD):
+        """
+        Args:
+            other (DatasetObject): parent object to find potential attachment links.
+            bypass_alignment_checking (bool): whether to bypass alignment checking when finding attachment links.
+            pos_thresh (float): position difference threshold to activate attachment, in meters.
+            orn_thresh (float): orientation difference threshold to activate attachment, in radians.
+
+        Returns:
+            RigidPrim or None: link belonging to @self.obj that should be aligned to that corresponding link of @other
+            RigidPrim or None: the corresponding link of @other
+        """
+        parent_candidates = self._get_parent_candidates(other)
+        if not parent_candidates:
+            return None, None
+
+        for child_link_name, parent_link_names in parent_candidates.items():
+            child_link = self.links[child_link_name]
+            for parent_link_name in parent_link_names:
+                parent_link = other.states[AttachedTo].links[parent_link_name]
+                if other.states[AttachedTo].children[parent_link_name] is None:
+                    pos_diff = np.linalg.norm(child_link.get_position() - parent_link.get_position())
+                    orn_diff = T.get_orientation_diff_in_radian(child_link.get_orientation(), parent_link.get_orientation())
+                    if bypass_alignment_checking or (pos_diff < pos_thresh and orn_diff < orn_thresh):
```

Comment:
for efficiency sake, should the bypass_alignment_checking check happen before we actually compute the pos / orn diff?

Reply:
done